students/johnn/project/dcd.py

```diff
@@ -15,187 +15,443 @@
+    # config.pub_interface = "tcp://{}:{}".format(config.myaddr,options.pub_port)
+    # # 
+    #config.myinterfaces = ( config.admin_interface, config.pub_interface )
 
-# define where to write the logs
-log_path = 'log/{}-{}.log'.format(scriptname, pid)
+    # choose logging level
+    if options.debug is None:
+        config.stream_log_level = logging.INFO
+    else:
+        config.stream_log_level = logging.DEBUG
 
-# define a stream and file handler
-log = logging.getLogger()
-log.setLevel(logging.DEBUG)
-formatter = logging.Formatter('%(asctime)s %(levelname)s %(module)s:%(threadName)s %(message)s')
 
-stream_handler = logging.StreamHandler()
-stream_handler.setLevel(options.stream_log_level)
-stream_handler.setFormatter(formatter)
+def setup_logging(config):
+    """
+    set up logging
 
-file_handler = logging.FileHandler(filename = log_path)
-file_handler.setLevel(logging.DEBUG)
-file_handler.setFormatter(formatter)
+    set basic logging, always logging to a file at level debug,
+    but logging to the screen at info or debug depending on the
+    presence of the --debug flag
+    """
 
-log.addHandler(file_handler)
-log.addHandler(stream_handler)
+    # define where to write the logs
+    log_path = 'log/{}-{}.log'.format(config.scriptname, config.pid)
 
-# write out startup info to the log
-log.info("process {}, PID {} starting".format(scriptname, pid))
-log.info("logging to {}".format(log_path))
-log.debug("screen log level {}".format(options.stream_log_level))
+    # define a stream and file handler
+    log = logging.getLogger()
+    # always write debug to the log
+    log.setLevel(logging.DEBUG)
+    formatter = logging.Formatter('%(asctime)s %(levelname)s %(module)s:%(threadName)s %(message)s')
 
-def decode_command(message):
+    # set up console output
+    stream_handler = logging.StreamHandler()
+    stream_handler.setLevel(config.stream_log_level)
+    stream_handler.setFormatter(formatter)
+
+    # set up file output
+    file_handler = logging.FileHandler(filename = log_path)
+    file_handler.setLevel(logging.DEBUG)
+    file_handler.setFormatter(formatter)
+
+    # start them up
+    log.addHandler(file_handler)
+    log.addHandler(stream_handler)
+
+    # write out startup info to the log
+    log.info("process {}, config.pid {} on {} starting".format(config.scriptname, config.pid, config.myaddr))
+    log.info("logging to {}".format(log_path))
+    log.debug("screen log level {}".format(config.stream_log_level))
+
+    return log
+
+
+def encode_message(command, key, value):
+    """
+    build a message string with the proper quoting
+    """
+    # we're building a string representation of a tuple,
+    # with each value quoted, like "('get', 'dog', '')"
+    return "({})".format(", ".join((repr(command), repr(key), repr(value))))
+
+
+def decode_message(message):
+    """
+    decode a message or return None
+    """
     try:
         command, key, value = eval(message)
```

Comment:
be very careful with using eval this way! if there is any chance a message could be intercepted and hacked!

Reply:
maybe make your messages JSON or something flexible but safer.